File: Serial app/Form1.vb
Imports System.IO.Ports
Imports System.Threading
Imports System.ComponentModel
Imports System.Text
Imports System.IO
Imports System

Public Class Form1

    Dim myPort As Array

    Public comOpen As Boolean

    Private readBuffer As String = String.Empty

    Dim stopMe As Boolean

    Delegate Sub SetTextCallback(ByVal [text] As String) 'Calling safe Thread

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        myPort = IO.Ports.SerialPort.GetPortNames() 'Geting list of all opened ports
        ComboBox1.Items.AddRange(myPort)
        Button1.Enabled = False
        Button2.Enabled = False
        Button3.Enabled = False


    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        SerialPort1.PortName = ComboBox1.Text
        SerialPort1.BaudRate = ComboBox2.Text
        Button2.Enabled = True
        Button3.Enabled = True
        TextBox1.Select()
        Try

            SerialPort1.Open() 'Connecting to Serialport
            SerialPort1.ReadTimeout = 10000
            comOpen = SerialPort1.IsOpen
            'Dim speech
            'speech = CreateObject("sapi.spvoice")
            'speech.speak("Connected to serialport Sucessful")

        Catch ex As Exception
            comOpen = False
            MsgBox("Error Open: " & ex.Message) 'Exception if Port is not open
        End Try
    End Sub

    Private Sub Button2_Click(sender As Object, e As EventArgs) Handles Button2.Click
        If TextBox1.Text = Nothing Then
            'Dim speech
            'speech = CreateObject("sapi.spvoice")
            'speech.speak("No text has been entered")
        End If
        SerialPort1.Write(TextBox2.Text) 'concatenate with \n
    End Sub

    Private Sub Button3_Click(sender As Object, e As EventArgs) Handles Button3.Click
        Button1.Enabled = True
        Button2.Enabled = False
        If MessageBox.Show("Do you really want to close the Serialport", "", MessageBoxButtons.YesNo) = Windows.Forms.DialogResult.No Then
            Try
                SerialPort1.DiscardInBuffer()
                SerialPort1.Close()
            Catch ex As IOException
                'Dim speech
                'speech = CreateObject("sapi.spvoice")
                'speech.speak("Serialport Closed")
            End Try
        Else
            Dim t As New Thread(AddressOf ClosePort)
            t.Start()
        End If
    End Sub
    Private Sub ClosePort()
        If SerialPort1.IsOpen Then SerialPort1.Close()
    End Sub

    Private Sub SerialPort1_ErrorReceived(sender As Object, e As SerialErrorReceivedEventArgs) Handles SerialPort1.ErrorReceived
        Try
            SerialPort1.Close()

        Catch ex As IOException

        End Try
    End Sub

    Private Sub ComboBox1_SelectedIndexChanged(sender As Object, e As EventArgs) Handles ComboBox1.SelectedIndexChanged
        Button1.Enabled = True
        ComboBox2.Text = "9600"
    End Sub


    Public Sub SerialPort1_DataReceived(ByVal sender As Object, ByVal e As System.IO.Ports.SerialDataReceivedEventArgs) Handles SerialPort1.DataReceived
        Try

            Dim i As Integer = SerialPort1.BytesToRead() 'Declearing Number of bytes to read

            If i >= 9 Then
                Dim data As Byte() = New Byte(i) {}

                'For Each item As Byte In data
                SerialPort1.Read(data, 0, i) 'Reading data from Serialport
                'If comOpen = True Then
                ReceivedText((data(0))) 'Sending the recieved data to another thread for displaying
                ReceivedText((data(1)))
                ReceivedText((data(2)))
                ReceivedText((data(3)))
                ReceivedText((data(4)))
                ReceivedText((data(5)))
                ReceivedText((data(6)))
                ReceivedText((data(7)))
                ReceivedText((data(8)))
                ReceivedText((data(9)))

            End If
            'Next
            'End If
        Catch ex As IOException
        Catch ex As IndexOutOfRangeException
        End Try
    End Sub

    Public Sub ReceivedText(ByVal [text] As String) 'displaying the data in text box using safe thread operation
        If Me.TextBox1.InvokeRequired Then
            Dim x As New SetTextCallback(AddressOf ReceivedText)
            Me.Invoke(x, New Object() {(text)})
        Else
            Me.TextBox1.Text &= [text] - 48 & vbCrLf  'append text

        End If

    End Sub

    Private Sub ComboBox1_KeyPress(sender As Object, e As KeyPressEventArgs) Handles ComboBox1.KeyPress
        If Asc(e.KeyChar) = 13 Then
            Button1.PerformClick()
        End If
    End Sub

    Private Sub TextBox1_KeyPress(sender As Object, e As KeyPressEventArgs) Handles TextBox1.KeyPress
        If Asc(e.KeyChar) = 13 Then
            SerialPort1.Write(TextBox2.Text)
        End If
    End Sub
End Class

File: Serial app/CS/Form1.cs
using System.Collections.Generic;
using System;
using System.Linq;
using System.Drawing;
using System.Diagnostics;
using System.Threading.Tasks;
using System.Data;
using System.Xml.Linq;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
using System.IO.Ports;
using System.Threading;
using System.ComponentModel;
using System.Text;
using System.IO;


namespace Serial_app
{
	public partial class Form1
	{
		public Form1()
		{
			InitializeComponent();
			
			//Added to support default instance behavour in C#
			if (defaultInstance == null)
				defaultInstance = this;
		}
		
#region Default Instance
		
		private static Form1 defaultInstance;
		
		/// <summary>
		/// Added by the VB.Net to C# Converter to support default instance behavour in C#
		/// </summary>
		public static Form1 Default
		{
			get
			{
				if (defaultInstance == null)
				{
					defaultInstance = new Form1();
					defaultInstance.FormClosed += new FormClosedEventHandler(defaultInstance_FormClosed);
				}
				
				return defaultInstance;
			}
		}
		
		static void defaultInstance_FormClosed(object sender, FormClosedEventArgs e)
		{
			defaultInstance = null;
		}
		
#endregion
		
		Array myPort;
		
		public bool comOpen;
		
		private string readBuffer = string.Empty;
		
		bool stopMe;
		
		delegate void SetTextCallback(string text); //Calling safe Thread
		
		public void Form1_Load(object sender, EventArgs e)
		{
			myPort = System.IO.Ports.SerialPort.GetPortNames(); //Geting list of all opened ports
			ComboBox1.Items.AddRange((object[]) myPort);
			Button1.Enabled = false;
			Button2.Enabled = false;
			Button3.Enabled = false;
			
			
		}
		
		public void Button1_Click(object sender, EventArgs e)
		{
			SerialPort1.PortName = ComboBox1.Text;
			SerialPort1.BaudRate = int.Parse(ComboBox2.Text);
			Button2.Enabled = true;
			Button3.Enabled = true;
			TextBox1.Select();
			try
			{
				
				SerialPort1.Open(); //Connecting to Serialport
				SerialPort1.ReadTimeout = 10000;
				comOpen = SerialPort1.IsOpen;
				//Dim speech
				//speech = CreateObject("sapi.spvoice")
				//speech.speak("Connected to serialport Sucessful")
				
			}
			catch (Exception ex)
			{
				comOpen = false;
				MessageBox.Show("Error Open: " + ex.Message); //Exception if Port is not open
			}
		}
		
		public void Button2_Click(object sender, EventArgs e)
		{
			if (TextBox1.Text == null)
			{
				//Dim speech
				//speech = CreateObject("sapi.spvoice")
				//speech.speak("No text has been entered")
			}
			SerialPort1.Write(TextBox2.Text); //concatenate with \n
		}
		
		public void Button3_Click(object sender, EventArgs e)
		{
			Button1.Enabled = true;
			Button2.Enabled = false;
			if (MessageBox.Show("Do you really want to close the Serialport", "", MessageBoxButtons.YesNo) == System.Windows.Forms.DialogResult.No)
			{
				try
				{
					SerialPort1.DiscardInBuffer();
					SerialPort1.Close();
				}
				catch (IOException)
				{
					//Dim speech
					//speech = CreateObject("sapi.spvoice")
					//speech.speak("Serialport Closed")
				}
			}
			else
			{
				Thread t = new Thread(new System.Threading.ThreadStart(ClosePort));
				t.Start();
			}
		}
		private void ClosePort()
		{
			if (SerialPort1.IsOpen)
			{
				SerialPort1.Close();
			}
		}
		
		public void SerialPort1_ErrorReceived(object sender, SerialErrorReceivedEventArgs e)
		{
			try
			{
				SerialPort1.Close();
				
			}
			catch (IOException)
			{
				
			}
		}
		
		public void ComboBox1_SelectedIndexChanged(object sender, EventArgs e)
		{
			Button1.Enabled = true;
			ComboBox2.Text = "9600";
		}
		
		
		public void SerialPort1_DataReceived(object sender, System.IO.Ports.SerialDataReceivedEventArgs e)
		{
			try
			{
				
				int i = SerialPort1.BytesToRead; //Declearing Number of bytes to read
				
				if (i >= 9)
				{
					byte[] data = new byte[i+ 1];
					
					//For Each item As Byte In data
					SerialPort1.Read(data, 0, i); //Reading data from Serialport
					//If comOpen = True Then
					ReceivedText(System.Convert.ToString((data[0]))); //Sending the recieved data to another thread for displaying
					ReceivedText(System.Convert.ToString((data[1])));
					ReceivedText(System.Convert.ToString((data[2])));
					ReceivedText(System.Convert.ToString((data[3])));
					ReceivedText(System.Convert.ToString((data[4])));
					ReceivedText(System.Convert.ToString((data[5])));
					ReceivedText(System.Convert.ToString((data[6])));
					ReceivedText(System.Convert.ToString((data[7])));
					ReceivedText(System.Convert.ToString((data[8])));
					ReceivedText(System.Convert.ToString((data[9])));
					
				}
				//Next
				//End If
			}
			catch (IOException)
			{
			}
			catch (IndexOutOfRangeException)
			{
			}
		}
		
		public void ReceivedText(string text) //displaying the data in text box using safe thread operation
		{
			if (this.TextBox1.InvokeRequired)
			{
				SetTextCallback x = new SetTextCallback(ReceivedText);
				this.Invoke(x, new object[] {(text)});
			}
			else
			{
				this.TextBox1.Text += int.Parse(text) - 48 + "\r\n" ; //append text
				
			}
			
		}
		
		public void ComboBox1_KeyPress(object sender, KeyPressEventArgs e)
		{
			if (Strings.Asc(e.KeyChar) == 13)
			{
				Button1.PerformClick();
			}
		}
		
		public void TextBox1_KeyPress(object sender, KeyPressEventArgs e)
		{
			if (Strings.Asc(e.KeyChar) == 13)
			{
				SerialPort1.Write(TextBox2.Text);
			}
		}
	}
	
}
